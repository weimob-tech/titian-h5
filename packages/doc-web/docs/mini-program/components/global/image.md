---
title: 图片
sidebar_custom_props:
  suffix: Image
hide_title: true
pagination_prev: null
pagination_next: null
side_iframe_path: '#/image'
---

# 图片 _Image_
**图片容器，在保留原生 img 的特性下，支持懒加载，自定义占位、加载失败等。**

import TabsLink from '@site/src/components/tabsLink';

<TabsLink id="ti-image-api" />

## 安装使用

```json showLineNumbers
{
  // 原生小程序
  "usingComponents": {
    "ti-image": "{{packageWeappName}}/image/index"
  },
  // titan-cli搭建的项目
  "usingComponents": {
    "ti-image": "platform://titian-mp/ti-image"
  }
}
```

## 用法示例

#### 基础用法

```html showLineNumbers
<ti-image width="180" height="180" src="{{src}}" />
```
#### 填充模式

```html showLineNumbers
<ti-image mode="contain" width="180" height="180" src="{{src}}" />
<ti-image mode="cover" width="180" height="180" src="{{src}}" />
<ti-image mode="fill" width="180" height="180" src="{{src}}" />
<ti-image mode="none" width="180" height="180" src="{{src}}" />
<ti-image mode="scaleDown" width="180" height="180" src="{{src}}" />
```

#### 自定义圆角
**use-global-style设置为true，可跟随全局圆角风格，即受控于css变量--base-radius-size**
```html showLineNumbers
<ti-image radius="18" width="100" height="100" src="{{src}}" />
<ti-image radius="100%" width="100" height="100" src="{{src}}" />
<ti-image use-global-style width="100" height="100" src="{{src}}" />
```
#### 懒加载

```html showLineNumbers
<ti-image lazy-load src="{{src}}" />
```
#### 图片比例（宽高比）
**aspect-ratio设置inherit，可跟随全局比例风格，即受控于css变量--image-aspect-ratio；也可以设置具体的宽高比。**
```html showLineNumbers
<ti-image aspect-ratio="2" src="{{src}}" />
<ti-image aspect-ratio="inherit" src="{{src}}" />
```
#### 加载中提示
**更多可用iconName详见[ti-icon组件](../icon#图标展示)**
```html showLineNumbers
<ti-image loading-icon="default-pic" src="{{src}}" />
<ti-image src="{{src}}">
  <div slot="loading" />
</ti-image>
```
#### 加载失败提示
**更多可用iconName详见[ti-icon组件](../icon#图标展示)**
```html showLineNumbers
<ti-image error-icon="default-pic" src="{{src}}" />
<ti-image src="{{src}}">
  <div slot="error" />
</ti-image>
```
## ti-image API
### 属性 **Properties**

| 名称            | 类型              | 必填 | 默认值        | 说明                                                                                 | 备注 |
| --------------- | ----------------- | ---- | ------------- | ------------------------------------------------------------------------------------ | ---- |
| src             | `string`          | 是   | -             | 图片资源地址                                                                         | -    |
| mode            | `string`          | 否   | `scaleToFill` | 图片填充模式，[可选类型](#图片填充模式mode)                                                                        | -    |
| width           | `string`          | 否   | -             | 宽度，默认单位为 px                                                                  | -    |
| height          | `string`          | 否   | -             | 高度，默认单位为 px                                                                  | -    |
| radius          | `string`          | 否   | -             | 圆角，默认单位为 px                                                                  | -    |
| show-loading     | `boolean`         | 否   | `false`       | 是否展示图片加载中提示                                                               | -    |
| show-error       | `boolean`         | 否   | `false`       | 是否展示图片加载失败提示                                                             | -    |
| loading-icon     | `string`          | 否   | `default-pic` | 加载中显示的图标                                                                     | -    |
| error-icon       | `string`          | 否   | `default-pic` | 加载失败显示的图标                                                                   | -    |
| loading-icon-size | `number`          | 否   | `72`          | 加载中显示的图标，尺寸                                                               | -    |
| error-icon-size   | `number`          | 否   | 72            | 加载失败显示的图标，尺寸                                                             | -    |
| use-loading-slot  | `boolean`         | 否   | false         | 使用 loading 插槽                                                                    | -    |
| use-error-slot    | `boolean`         | 否   | false         | 使用 error 插槽                                                                      | -    |
| lazy-load        | `boolean`         | 否   | false         | 图片懒加载，在即将进入一定范围（上下三屏）时才开始加载                               | -    |
| use-global-style  | `boolean`         | 否   | false         | 设为true，可跟随全局圆角风格，即受控于css变量--base-radius-size                      | -    |
| aspect-ratio     | `string` \| `number` | 否   | -1            | 宽高比，如果设置为`inherit`，可跟随全局比例风格，即受控于css变量--image-aspect-ratio | -    |
| ext-style        | `string` \| `Record<string, string>`          | 否   | -             | 根节点样式                                                                           | -    |

#### 图片填充模式（mode）

| 名称         | 含义                                                                                    |
| ------------ | --------------------------------------------------------------------------------------- |
| contain      | 同 aspectFit， 保持宽高缩放图片，使图片的长边能完全显示出来                             |
| cover        | 同 aspectFill， 保持宽高缩放图片，使图片的短边能完全显示出来，裁剪长边                  |
| fill         | 同 scaleToFill，拉伸图片，使图片填满元素                                                |
| none         | 同 center，保持图片原有尺寸                                                             |
| scale-down    | 内容的尺寸与 none 或 contain 中的一个相同，取决于它们两个之间谁得到的对象尺寸会更小一些 |
| top          | 裁剪模式，不缩放图片，只显示图片的顶部区域                                              |
| bottom       | 裁剪模式，不缩放图片，只显示图片的底部区域                                              |
| center       | 裁剪模式，不缩放图片，只显示图片的中间区域                                              |
| left         | 裁剪模式，不缩放图片，只显示图片的左边区域                                              |
| right        | 裁剪模式，不缩放图片，只显示图片的右边区域                                              |
| top left     | 裁剪模式，不缩放图片，只显示图片的左上边区域                                            |
| top right    | 裁剪模式，不缩放图片，只显示图片的右上边区域                                            |
| bottom left  | 裁剪模式，不缩放图片，只显示图片的左下边区域                                            |
| bottom right | 裁剪模式，不缩放图片，只显示图片的右下边区域                                            |

### 事件 **Events**

| 名称    | 参数列表                                                 | 描述                 | 备注 |
| ------- | -------------------------------------------------------- | -------------------- | ---- |
| bind:load  |  `(e: WechatMiniprogram.CustomEvent<{width: number, height: number}>) => void` | 当图片载入完毕时触发 | -    |
| bind:error | `(e: WechatMiniprogram.CustomEvent) => void`                                                        | 当错误发生时触发     | -    |

### 插槽 **Slots**

| 名称    | 说明         | 备注 |
| ------- | ------------ | ---- |
| loading | 加载中展示   | -    |
| error   | 加载失败展示 | -    |

### 外部样式类 **External Classes**

| 名称     | 说明         | 备注 |
| -------- | ------------ | ---- |
| ext-class | 根节点样式类 | -    |
### CSS 变量 **CSS Variable**
| 变量                 | 默认值 | 说明                                                                       | 备注 |
| -------------------- | ------ | -------------------------------------------------------------------------- | ---- |
| --image-aspect-ratio | `1`    | 图片比例，aspect-ratio属性为inherit下生效，一般设置在项目根节点，作用于全局 | -    |
| --base-radius-size   | `0rpx`  | 图片圆角，use-global-style属性为true下生效，一般设置在项目根节点，作用于全局 | -    |