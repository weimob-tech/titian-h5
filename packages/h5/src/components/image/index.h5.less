@import '../common/static/utils.less';
@import '../common/static/var.less';
@import '../common/static/mixins/index.less';

:host {
  display: contents;
}
@image:~ '@{project}-image';

.@{image} {
  display: inline-block;
  position: relative;
  z-index: 0;
  width: 100%;
  overflow: hidden;

  outline: none;

  vertical-align: middle;

  &-ratio::after {
    content: '';

    display: block;
    padding-bottom: calc(1 / (@image-aspect-ratio) * 100%);
  }
  @supports (aspect-ratio: 1) {
    &-ratio {
      aspect-ratio: @image-aspect-ratio;

      &::after {
        content: none;
      }
    }
  }

  &-notice {
    position: absolute;
    z-index: -1;
    top: 50%;
    left: 50%;

    transform: translate(-50%, -50%);

    color: @neutral-color-5;
    font-size: @T28;
    white-space: nowrap;

    --loading-color: @neutral-color-5;
  }

  &-img {
    vertical-align: middle;

    &-ratio {
      position: absolute;
      top: 0;
      left: 0;
      width: 100%;
      height: 100%;
    }

    &-fill {
      width: 100%;
      height: 100%;
    }

    &-contain {
      object-fit: contain;

      width: 100%;
      height: 100%;
    }

    &-cover {
      object-fit: cover;

      width: 100%;
      height: 100%;
    }

    &-scaledown {
      object-fit: scale-down;

      width: 100%;
      height: 100%;
    }

    &-center {
      position: absolute;
      top: 50%;
      left: 50%;

      transform: translate(-50%, -50%);
    }

    &-top {
      position: absolute;
      top: 0;
      left: 50%;

      transform: translateX(-50%);
    }

    &-bottom {
      position: absolute;
      bottom: 0;
      left: 50%;

      transform: translateX(-50%);
    }

    &-left {
      position: absolute;
      top: 50%;
      left: 0;

      transform: translateY(-50%);
    }

    &-right {
      position: absolute;
      top: 50%;
      right: 0;

      transform: translateY(-50%);
    }

    &-topleft {
      position: absolute;
      top: 0;
      left: 0;
    }

    &-topright {
      position: absolute;
      top: 0;
      right: 0;
    }

    &-bottomleft {
      position: absolute;
      bottom: 0;
      left: 0;
    }

    &-bottomright {
      position: absolute;
      right: 0;
      bottom: 0;
    }

    &-widthfix {
      width: 100%;
      height: auto;
    }

    &-heightfix {
      width: auto;
      height: 100%;
    }
  }
}
