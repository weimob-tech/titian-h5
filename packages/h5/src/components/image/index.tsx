import { Component, Prop, Element, State, h, Listen, Watch, Event, EventEmitter } from '@stencil/core';
import { JSXBase } from '@stencil/core/internal';
import { BasicComponentAbstract } from '../common/basic/BasicComponent';
import { stringToAttrStyle, throttle, addShadowRootStyle } from '../common/utils';
import * as namespace from '../common/utils/namespace';
import addUnit from '../common/utils/suffix';

enum ImageModeEnum {
  'SCALE_TO_FILL' = 'scaleToFill',
  'ASPECT_FIT' = 'aspectFit',
  'ASPECT_FILL' = 'aspectFill',
  'WIDTH_FIX' = 'widthFix',
  'HEIGHT_FIX' = 'heightFix',
  'TOP' = 'top',
  'BOTTOM' = 'bottom',
  'CENTER' = 'center',
  'LEFT' = 'left',
  'RIGHT' = 'right',
  'TOP_LEFT' = 'top left',
  'TOP_RIGHT' = 'top right',
  'BOTTOM_LEFT' = 'bottom left',
  'BOTTOM_RIGHT' = 'bottom right',
  'SCALE_DOWN' = 'scaleDown',
  'NONE' = 'none',
  'FILL' = 'fill',
  'COVER' = 'cover',
  'CONTAIN' = 'contain',
}
type MainStyleAttrs = {
  'width'?: string;
  'height'?: string;
  'border-radius'?: string;
};
@Component({
  tag: 'ti-image',
  styleUrls: {
    pc: 'index.pc.less',
    h5: 'index.h5.less',
  },
  shadow: true,
})
export class TiImage implements BasicComponentAbstract {
  @Element() host: HTMLElement;

  @Prop() extCss = '';

  @Prop() extClass?: string = '';

  @Prop() extStyle?: string | JSXBase.HTMLAttributes<Record<string, unknown>>['style'];

  @Prop() src: string;

  @Prop() alt: string;

  @Prop() width?: string | number;

  @Prop() height?: string | number;

  @Prop() radius?: string | number;

  @Prop() mode?: ImageModeEnum = ImageModeEnum.SCALE_TO_FILL;

  @Prop() lazyLoad?: boolean;

  @Prop() showLoading?: boolean = true;

  @Prop() showError?: boolean = true;

  @Prop() useLoadingSlot?: boolean;

  @Prop() useErrorSlot?: boolean;

  @Prop() loadingIcon?: string = 'default-pic';

  @Prop() errorIcon?: string = 'default-pic';

  @Prop() loadingIconSize?: number = 72;

  @Prop() errorIconSize?: number = 72;

  @Prop() useGlobalStyle?: boolean;

  @Prop() aspectRatio?: number | 'inherit' = -1;

  @State() loading = true;

  @State() isError = false;

  @State() imageWidth: number;

  @State() imageHeight: number;

  @State() show = false;

  @State() mainStyle: MainStyleAttrs = {};

  @State() imgStyle = {};

  @State() modeClass = {};

  @Event({ bubbles: false, composed: false }) load: EventEmitter<{ width: number; height: number }>;

  @Event({ bubbles: false, composed: false }) error: EventEmitter<never>;

  private handleLazyLoad: any;

  private imgBoxRef = null;

  private imgRef = null;

  private isSupportLoading = false;

  private timer = null;

  @Listen('scroll', { target: 'window', capture: true })
  handleScroll() {
    if (!this.lazyLoad || !this.loading || !this.handleLazyLoad) return;
    this.handleLazyLoad();
  }

  @Watch('radius')
  @Watch('width')
  @Watch('height')
  @Watch('aspectRatio')
  @Watch('extStyle')
  getMainStyle() {
    const { width, height, radius, extStyle, useGlobalStyle, aspectRatio } = this;
    const mainStyle: MainStyleAttrs = {};
    if (useGlobalStyle) {
      const curRadius = radius ? addUnit(radius) : '0px';
      mainStyle['border-radius'] = `calc(var(--base-radius-size, 0px) + ${curRadius})`;
    } else {
      mainStyle['border-radius'] = addUnit(radius);
    }
    if (aspectRatio > 0) {
      mainStyle['--image-aspect-ratio'] = aspectRatio;
    }

    if (width) {
      mainStyle.width = addUnit(width);
    }
    if (height) {
      mainStyle.height = addUnit(height);
      // 没有宽度，只有高度（且不是百分比）
      const supportsAspect = CSS.supports('aspect-ratio: 1');
      const useAspectRatio = aspectRatio > 0 || aspectRatio === 'inherit';
      if (
        !supportsAspect &&
        useAspectRatio &&
        !width &&
        (typeof height === 'string' ? height.indexOf('%') === -1 : true)
      ) {
        mainStyle.width = `calc(${addUnit(height)} * var(--image-aspect-ratio, 1))`;
      }
    }

    this.mainStyle = Object.assign(mainStyle, stringToAttrStyle(extStyle));
  }

  @Watch('mode')
  getImgStyle() {
    const { mode } = this;
    const MODE_MAP = {
      none: 'center',
      scaleToFill: 'fill',
      aspectFill: 'cover',
      aspectFit: 'contain',
    };
    const currentMode = MODE_MAP[mode] || mode || ImageModeEnum.SCALE_TO_FILL;
    this.modeClass = currentMode.toLowerCase().replace(/\s/g, '');
  }

  @Watch('src')
  update(val, oldVal) {
    if (val !== oldVal) {
      this.loading = true;
      this.isError = false;
      this.LazyLoadImg();
    }
  }

  onLoad = () => {
    this.isError = false;
    this.loading = false;
    this.handleLazyLoad = null;
    this.load.emit({ width: this.imgRef.naturalWidth, height: this.imgRef.naturalHeight });
  };

  onError = () => {
    this.isError = true;
    this.loading = false;
    this.error.emit();
  };

  componentWillLoad() {
    this.isSupportLoading = 'loading' in document.createElement('img');
    this.getMainStyle();
    this.getImgStyle();
    addShadowRootStyle.call(this);
  }

  LazyLoadImg() {
    if (!this.lazyLoad || this.isSupportLoading) return;
    const isOutScreen = () => {
      const { top, bottom } = this.imgBoxRef.getBoundingClientRect();
      return top > window.innerHeight * 3 || bottom < window.innerHeight * -3;
    };
    this.handleLazyLoad = throttle(() => {
      if (isOutScreen()) return;
      this.imgRef.src = this.src;
    });
    this.handleLazyLoad();
  }

  onTouchStart = () => {
    const img = this.host.querySelector('img');
    if (img) {
      img.src = this.src;
      return;
    }
    this.timer = setTimeout(() => {
      this.longpress();
    }, 500);
  };

  onTouchMove = () => {
    clearTimeout(this.timer);
  };

  onTouchEnd = () => {
    clearTimeout(this.timer);
  };

  longpress = () => {
    const img = document.createElement('img');
    img.src = this.src;
    img.setAttribute('style', 'opacity: 0;position:absolute;top:0;left:0;width:100%;height:100%');
    this.host.appendChild(img);
  };

  componentDidLoad() {
    this.LazyLoadImg();
  }

  render() {
    const {
      src,
      alt,
      extClass,
      mainStyle,
      imgStyle,
      loading,
      isError,
      modeClass,
      showLoading,
      useLoadingSlot,
      loadingIcon,
      loadingIconSize,
      useErrorSlot,
      showError,
      errorIcon,
      errorIconSize,
      isSupportLoading,
      lazyLoad,
      aspectRatio,
    } = this;
    const useAspectRatio = aspectRatio > 0 || aspectRatio === 'inherit';
    const imgClass = `${namespace.join('image-img', [modeClass, { ratio: useAspectRatio }])}`;
    const iconClass = `${namespace.join('image-notice')}`;
    return (
      <div
        class={`${namespace.join('image', [{ ratio: useAspectRatio }])} ${extClass}`}
        style={mainStyle}
        part={extClass}
        ref={el => el && (this.imgBoxRef = el)}
        aria-hidden="true"
        onTouchStart={this.onTouchStart}
        onTouchMove={this.onTouchMove}
        onTouchEnd={this.onTouchEnd}
      >
        {!isError &&
          (isSupportLoading || !lazyLoad ? (
            <img
              ref={el => el && (this.imgRef = el)}
              class={imgClass}
              src={src}
              alt={alt}
              style={imgStyle}
              loading={lazyLoad ? 'lazy' : 'eager'}
              onLoad={this.onLoad}
              onError={this.onError}
            />
          ) : (
            <img
              ref={el => el && (this.imgRef = el)}
              class={imgClass}
              date-src={src}
              alt={alt}
              style={imgStyle}
              onLoad={this.onLoad}
              onError={this.onError}
            />
          ))}
        {showLoading && loading && (
          <div class={iconClass}>
            {useLoadingSlot ? <slot name="loading" /> : <ti-icon name={loadingIcon} size={loadingIconSize} />}
          </div>
        )}
        {showError && isError && (
          <div class={iconClass}>
            {useErrorSlot ? <slot name="error" /> : <ti-icon name={errorIcon} size={errorIconSize} />}
          </div>
        )}
        <slot />
      </div>
    );
  }
}
