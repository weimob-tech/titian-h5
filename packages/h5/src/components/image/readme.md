# ti-image



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute           | Description      | Type                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    | Default                       |
| ----------------- | ------------------- | ---------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------- |
| `alt`             | `alt`               |                  | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `undefined`                   |
| `errorIcon`       | `error-icon`        |                  | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `'default-pic'`               |
| `errorIconSize`   | `error-icon-size`   |                  | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `72`                          |
| `extClass`        | `ext-class`         | 额外的类名，添加到根节点的元素上 | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `''`                          |
| `extStyle`        | `ext-style`         | 额外的样式            | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `undefined`                   |
| `height`          | `height`            |                  | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `240`                         |
| `lazyLoad`        | `lazy-load`         |                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `undefined`                   |
| `loadingIcon`     | `loading-icon`      |                  | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `'default-pic'`               |
| `loadingIconSize` | `loading-icon-size` |                  | `number`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `72`                          |
| `mode`            | `mode`              |                  | `ImageModeEnum.ASPECT_FILL \| ImageModeEnum.ASPECT_FIT \| ImageModeEnum.BOTTOM \| ImageModeEnum.BOTTOM_LEFT \| ImageModeEnum.BOTTOM_RIGHT \| ImageModeEnum.CENTER \| ImageModeEnum.CONTAIN \| ImageModeEnum.COVER \| ImageModeEnum.FILL \| ImageModeEnum.HEIGHT_FIX \| ImageModeEnum.LEFT \| ImageModeEnum.NONE \| ImageModeEnum.RIGHT \| ImageModeEnum.SCALE_DOWN \| ImageModeEnum.SCALE_TO_FILL \| ImageModeEnum.TOP \| ImageModeEnum.TOP_LEFT \| ImageModeEnum.TOP_RIGHT \| ImageModeEnum.WIDTH_FIX` | `ImageModeEnum.SCALE_TO_FILL` |
| `radius`          | `radius`            |                  | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `undefined`                   |
| `showError`       | `show-error`        |                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `true`                        |
| `showLoading`     | `show-loading`      |                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `true`                        |
| `src`             | `src`               |                  | `string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | `undefined`                   |
| `useErrorSlot`    | `use-error-slot`    |                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `undefined`                   |
| `useLoadingSlot`  | `use-loading-slot`  |                  | `boolean`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               | `undefined`                   |
| `width`           | `width`             |                  | `number \| string`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | `320`                         |


## Events

| Event   | Description | Type                                              |
| ------- | ----------- | ------------------------------------------------- |
| `click` |             | `CustomEvent<never>`                              |
| `error` |             | `CustomEvent<never>`                              |
| `load`  |             | `CustomEvent<{ width: number; height: number; }>` |


## Dependencies

### Depends on

- [ti-icon](../icon)

### Graph
```mermaid
graph TD;
  ti-image --> ti-icon
  style ti-image fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
