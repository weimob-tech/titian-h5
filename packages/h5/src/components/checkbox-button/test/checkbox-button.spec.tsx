import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TiCheckboxButton } from '../index';
import { TiImage } from '../../image';
import { TestContainer } from '../../common/test/container';

describe('渲染 checkbox button', () => {
  it('基本渲染', async () => {
    const page = await newSpecPage({
      components: [TiCheckboxButton, TestContainer],
      template: () => (
        <test-container>
          <ti-checkbox-button label="复选按钮文字" />
          <ti-checkbox-button>复选按钮文字</ti-checkbox-button>
        </test-container>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('渲染默认选中状态', async () => {
    const page = await newSpecPage({
      components: [TiCheckboxButton, TestContainer],
      template: () => (
        <test-container>
          <ti-checkbox-button label="默认选中" checked />
          <ti-checkbox-button label="默认选中" default-checked />
        </test-container>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('渲染禁用按钮', async () => {
    const page = await newSpecPage({
      components: [TiCheckboxButton],
      template: () => <ti-checkbox-button label="禁用" disabled />,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('修改按钮颜色', async () => {
    const page = await newSpecPage({
      components: [TiCheckboxButton, TestContainer],
      template: () => (
        <test-container>
          <ti-checkbox-button color="#ffff00" label="颜色" />
          <ti-checkbox-button color="rgb(0,0,0)" label="颜色" />
          <ti-checkbox-button color="rgba(0,0,0,0.5)" label="颜色" />
        </test-container>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('渲染组件内容', async () => {
    const page = await newSpecPage({
      components: [TiCheckboxButton, TiImage, TestContainer],
      template: () => (
        <test-container>
          <ti-checkbox-button label="设置右侧图标" icon="sort-inactive" />
          <ti-checkbox-button label="自定义左侧内容">
            <ti-image
              ext-style="margin-right: 6rpx"
              slot="prefix"
              width="40"
              radius="0"
              height="40"
              src="https://image-c-dev.weimobwmc.com/qa-On6X/8b97cd488593474ba4a8ccaa3c1a493f.png"
            />
          </ti-checkbox-button>
          <ti-checkbox-button label="自定义右侧内容">
            <ti-image
              ext-style="margin-left: 6rpx"
              slot="suffix"
              width="40"
              radius="0"
              height="40"
              src="https://image-c-dev.weimobwmc.com/qa-On6X/8b97cd488593474ba4a8ccaa3c1a493f.png"
            />
          </ti-checkbox-button>
        </test-container>
      ),
    });
    expect(page.root).toMatchSnapshot();
  });

  it('监听点击事件', async () => {
    let checked = false;
    const change = jest.fn((e: CustomEvent) => {
      checked = e.detail;
    });
    const page = await newSpecPage({
      components: [TiCheckboxButton],
      template: () => <ti-checkbox-button label="单选按钮" onChange={change} />,
    });

    {
      // 点击
      const tag = page.root.shadowRoot.querySelector<HTMLDivElement>('ti-tag');
      tag.click();
      expect(checked).toBeTruthy();
      tag.click();
      expect(checked).toBeFalsy();
    }
  });

  it('禁用情况下，监听点击事件', async () => {
    const change = jest.fn();
    const page = await newSpecPage({
      components: [TiCheckboxButton],
      template: () => <ti-checkbox-button label="单选按钮" disabled onChange={change} />,
    });

    {
      // 点击
      const tag = page.root.shadowRoot.querySelector<HTMLDivElement>('ti-tag');
      tag.click();
      expect(change).not.toBeCalled();
    }
  });
});
