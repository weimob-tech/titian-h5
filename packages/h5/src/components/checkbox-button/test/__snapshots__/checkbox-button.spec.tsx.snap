// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`渲染 checkbox button 修改按钮颜色 1`] = `
<test-container>
  <!---->
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button" style="--checkbox-button-disabled-border-color: rgba(255, 255, 0, 0.2); --checkbox-button-disabled-text-color: rgba(255, 255, 0, 0.4); --checkbox-button-checked-bg-color: rgba(255, 255, 0, 0.1); --checkbox-button-checked-text-color: rgba(255, 255, 0, 1); --checkbox-button-checked-border-color: rgba(255, 255, 0, 0.4);">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          颜色
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button" style="--checkbox-button-disabled-border-color: rgba(0,0,0,0.2); --checkbox-button-disabled-text-color: rgba(0,0,0,0.4); --checkbox-button-checked-bg-color: rgba(0,0,0,0.1); --checkbox-button-checked-text-color: rgba(0,0,0,1); --checkbox-button-checked-border-color: rgba(0,0,0,0.4);">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          颜色
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button" style="--checkbox-button-disabled-border-color: rgba(0,0,0,0.2); --checkbox-button-disabled-text-color: rgba(0,0,0,0.4); --checkbox-button-checked-bg-color: rgba(0,0,0,0.1); --checkbox-button-checked-text-color: rgba(0,0,0,1); --checkbox-button-checked-border-color: rgba(0,0,0,0.4);">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          颜色
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
</test-container>
`;

exports[`渲染 checkbox button 基本渲染 1`] = `
<test-container>
  <!---->
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          复选按钮文字
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          <slot></slot>
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
    复选按钮文字
  </ti-checkbox-button>
</test-container>
`;

exports[`渲染 checkbox button 渲染禁用按钮 1`] = `
<ti-checkbox-button>
  <mock:shadow-root>
    <div aria-hidden="true" class="titian-checkbox-button titian-checkbox-button-disabled">
      <ti-tag lefticon="" righticon="">
        <slot name="prefix" slot="prefix"></slot>
        禁用
        <slot name="suffix" slot="suffix"></slot>
      </ti-tag>
    </div>
  </mock:shadow-root>
</ti-checkbox-button>
`;

exports[`渲染 checkbox button 渲染组件内容 1`] = `
<test-container>
  <!---->
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button">
        <ti-tag lefticon="" righticon="sort-inactive">
          <slot name="prefix" slot="prefix"></slot>
          设置右侧图标
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          自定义左侧内容
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
    <ti-image ext-style="margin-right: 6rpx" slot="prefix">
      <mock:shadow-root>
        <div aria-hidden="true" class="titian-image" part="" style="border-radius: 0px; width: 20px; height: 20px; margin-right:  6rpx;">
          <img class="titian-image-img titian-image-img-fill" loading="eager" src="https://image-c-dev.weimobwmc.com/qa-On6X/8b97cd488593474ba4a8ccaa3c1a493f.png">
          <div class="titian-image-notice">
            <ti-icon name="default-pic" size="72"></ti-icon>
          </div>
          <slot></slot>
        </div>
      </mock:shadow-root>
    </ti-image>
  </ti-checkbox-button>
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          自定义右侧内容
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
    <ti-image ext-style="margin-left: 6rpx" slot="suffix">
      <mock:shadow-root>
        <div aria-hidden="true" class="titian-image" part="" style="border-radius: 0px; width: 20px; height: 20px; margin-left:  6rpx;">
          <img class="titian-image-img titian-image-img-fill" loading="eager" src="https://image-c-dev.weimobwmc.com/qa-On6X/8b97cd488593474ba4a8ccaa3c1a493f.png">
          <div class="titian-image-notice">
            <ti-icon name="default-pic" size="72"></ti-icon>
          </div>
          <slot></slot>
        </div>
      </mock:shadow-root>
    </ti-image>
  </ti-checkbox-button>
</test-container>
`;

exports[`渲染 checkbox button 渲染默认选中状态 1`] = `
<test-container>
  <!---->
  <ti-checkbox-button>
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button titian-checkbox-button-checked">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          默认选中
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
  <ti-checkbox-button default-checked="">
    <mock:shadow-root>
      <div aria-hidden="true" class="titian-checkbox-button titian-checkbox-button-checked">
        <ti-tag lefticon="" righticon="">
          <slot name="prefix" slot="prefix"></slot>
          默认选中
          <slot name="suffix" slot="suffix"></slot>
        </ti-tag>
      </div>
    </mock:shadow-root>
  </ti-checkbox-button>
</test-container>
`;
